compiler-core/src/analyse/name.rs

```diff
@@ -0,0 +1,116 @@
+use ecow::EcoString;
+
+use crate::ast::{ArgNames, SrcSpan};
+
+use super::{BadNameKind, Error};
+use heck::{ToSnakeCase, ToUpperCamelCase};
+
+type BadName = (SrcSpan, EcoString);
+
+pub fn check_valid_name(
+    location: SrcSpan,
+    name: &EcoString,
+    kind: BadNameKind,
+) -> Option<(Error, BadName)> {
+    check_bad_name(location, name, kind)
+        .map(|error| (error, (location, EcoString::from(name.to_snake_case()))))
+}
+
+pub fn check_valid_discard_name(location: SrcSpan, name: &EcoString) -> Option<(Error, BadName)> {
+    check_bad_name(location, name, BadNameKind::Discard).map(|error| {
+        (
+            error,
+            (
+                location,
+                EcoString::from(format!("_{}", name.to_snake_case())),
```

Comment:
Favour `eco_format!` rather than building a string and then converting it to an ecostring